File: src/main/java/net/infstudio/goki/api/stat/StatBase.scala
package net.infstudio.goki.api.stat

import java.util
import java.util.ArrayList

import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.client.resources.I18n
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.item.ItemStack
import net.minecraft.world.World
import net.minecraftforge.common.ForgeHooks

class StatBase extends Stat {
    val stats = new util.ArrayList[StatBase](32)
    var totalStats = 0
    //	 public static final StatBase STAT_FOCUS = new StatFocus(14, "grpg_Focus",
    //	 "Focus", 25);
    var imageID = 0
    var key: String = null
    var costMultiplier = 1.0F
    var limitMultiplier = 1.0F
    var bonusMultiplier = 1.0F
    var enabled = true
    var name: String = null
    val des: String = null
    private var limit = 0

    def this() {
        this()
        this.imageID = -1
        this.limit = 0
        this.key = "Dummy"
        this.name = "Dummy StatBase"
        stats.add(this)
    }

    def this(imgId: Int, key: String, limit: Int) {
        this()
        this.imageID = imgId
        this.limit = limit
        this.key = key
        stats.add(this)
        totalStats += 1
    }

    protected def getFinalBonus(currentBonus: Float): Float = currentBonus * GokiConfig.globalModifiers.globalBonusMultiplier

    override def getKey: String = key

    override def getBonus(player: EntityPlayer): Float = getBonus(DataHelper.getPlayerStatLevel(player, this)) * bonusMultiplier

    override def getAppliedDescriptionVar(player: EntityPlayer): Array[Float] = Array[Float](DataHelper.trimDecimals(getBonus(player) * 100, 1))

    override def getCost(level: Int): Int = ((Math.pow(level, 1.6D) + 6.0D + level) * GokiConfig.globalModifiers.globalCostMultiplier).toInt

    override def needAffectedByStat(obj: Any*): Boolean = {
        if (obj(1).isInstanceOf[ItemStack] && obj(2).isInstanceOf[BlockPos] && obj(3).isInstanceOf[World]) {
            val stack = obj(1).asInstanceOf[ItemStack]
            val pos = obj(2).asInstanceOf[BlockPos]
            val world = obj(3).asInstanceOf[World]
            return ForgeHooks.isToolEffective(world, pos, stack)
        }
        false
    }

    override def getLimit: Int = {
        if (GokiConfig.globalModifiers.globalLimitMultiplier <= 0.0F) return 127
        (this.limit * GokiConfig.globalModifiers.globalLimitMultiplier).toInt
    }

    override def getAppliedBonus(player: EntityPlayer, `object`: Any): Float = if (needAffectedByStat(`object`)) getBonus(player)
    else 0

    protected def getPlayerStatLevel(player: EntityPlayer): Int = DataHelper.getPlayerStatLevel(player, this)

    def needAffectedByStat(object1: Any, object2: Any, object3: Any): Boolean = {
        if (object1.isInstanceOf[ItemStack] && object2.isInstanceOf[BlockPos] && object3.isInstanceOf[World]) {
            val stack = object1.asInstanceOf[ItemStack]
            val pos = object2.asInstanceOf[BlockPos]
            val world = object3.asInstanceOf[World]
            if (ForgeHooks.isToolEffective(world, pos, stack)) return true
        }
        needAffectedByStat(object1)
    }

    @SideOnly(Side.CLIENT) def getLocalizedName: String = I18n.format(this.key + ".name")

    @SideOnly(Side.CLIENT) def getLocalizedDes(player: EntityPlayer): String = I18n.format(this.key + ".des", this.getAppliedDescriptionVar(player)(0))
}

File: src/main/java/net/infstudio/goki/api/stat/StatBase.java
package net.infstudio.goki.api.stat;

import it.unimi.dsi.fastutil.objects.ObjectArrayList;
import it.unimi.dsi.fastutil.objects.ObjectList;
import net.infstudio.goki.common.config.Configurable;
import net.infstudio.goki.common.config.GokiConfig;
import net.infstudio.goki.common.config.stats.StatConfig;
import net.infstudio.goki.common.utils.DataHelper;
import net.infstudio.goki.common.utils.Reference;
import net.minecraft.client.resources.language.I18n;
import net.minecraft.core.BlockPos;
import net.minecraft.resources.ResourceLocation;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.common.ForgeConfigSpec;
import net.minecraftforge.common.util.LazyOptional;
import net.minecraftforge.registries.ForgeRegistryEntry;
import net.minecraftforge.registries.IForgeRegistry;
import net.minecraftforge.registries.RegistryBuilder;

import java.util.Objects;

public abstract class StatBase<T extends StatConfig> extends ForgeRegistryEntry<Stat> implements Stat, Configurable<T> {
    public static final IForgeRegistry<Stat> REGISTRY = new RegistryBuilder<Stat>()
            .setName(new ResourceLocation(Reference.MODID, "stats"))
            .setType(Stat.class)
            .create();
    public static final ObjectList<StatBase<?>> stats = new ObjectArrayList<>(16);
    public static LazyOptional<Integer> totalStats = LazyOptional.of(() -> REGISTRY.getValues().size());
    public int imageID;
    public String key;

    private boolean enabled = true;
    private final int limit;

    public final T config;
    public final ForgeConfigSpec configSpec;

    public StatBase(int imgId, String key, int limit) {
        this.imageID = imgId;
        this.limit = limit;
        this.key = key;
        stats.add(this);

        final var specPair = new ForgeConfigSpec.Builder()
                .configure(this::createConfig);
        configSpec = specPair.getRight();
        config = specPair.getLeft();

        setRegistryName(Reference.MODID, key.toLowerCase());
        REGISTRY.register(this);
    }

    protected float getFinalBonus(float currentBonus) {
        return (float) (currentBonus * config.bonusMultiplier.get() * GokiConfig.SERVER.globalBonusMultiplier.get());
    }

    @Override
    @SuppressWarnings("unchecked")
    public T createConfig(ForgeConfigSpec.Builder builder) {
        builder.comment("Server configuration settings")
                .push("stats." + key);
        return (T) new StatConfig(builder);
    }

    @Override
    public double getBonus(Player player) {
        return getBonus(DataHelper.getPlayerStatLevel(player, this)) * getBonusMultiplier();
    }

    @Override
    public double[] getDescriptionFormatArguments(Player player) {
        return new double[]
                {DataHelper.trimDecimals(getBonus(player) * 100, 1)};
    }

    @Override
    public int getCost(int level) {
        return (int) ((Math.pow(level, 1.6D) + 6.0D + level) * config.costMultiplier.get() * GokiConfig.SERVER.globalCostMultiplier.get());
    }

    @Override
    public boolean isEffectiveOn(Object... obj) {
        if (obj[0] instanceof ItemStack stack && obj[1] instanceof BlockPos pos && obj[2] instanceof Level world) {

            return stack.isCorrectToolForDrops(world.getBlockState(pos));
        }
        return false;
    }

    @Override
    public int getLimit() {
        var limit = this.limit;
        if (config.maxLevel.get() > 0) return config.maxLevel.get();
        if (GokiConfig.SERVER.globalLimitMultiplier.get() <= 0) {
            return limit;
        }
        return (int) (limit * GokiConfig.SERVER.globalLimitMultiplier.get());
    }

    @Override
    public double getAppliedBonus(Player player, Object object) {
        if (isEffectiveOn(object))
            return getBonus(player);
        else
            return 0;
    }

    protected final int getPlayerStatLevel(Player player) {
        return DataHelper.getPlayerStatLevel(player, this);
    }

    public final boolean isEffectiveOn(ItemStack stack, BlockPos pos, Level world) {
        if (stack.isCorrectToolForDrops(world.getBlockState(pos)))
            return true;
        else return isEffectiveOn(stack);
    }

    @OnlyIn(Dist.CLIENT)
    public String getLocalizedName() {
        return I18n.get("skill.gokistats." + this.key);
    }

    @OnlyIn(Dist.CLIENT)
    public String getLocalizedDescription(Player player) {
        return I18n.get("skill.gokistats." + this.key + ".text",
                this.getDescriptionFormatArguments(player)[0]);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof StatBase)) return false;
        var statBase = (StatBase<?>) o;
        return Objects.equals(getRegistryName(), statBase.getRegistryName());
    }

    @Override
    public int hashCode() {
        return Objects.hash(getRegistryName());
    }

    @Override
    public String toString() {
        return getRegistryName().toString();
    }

    @Override
    public boolean isEnabled() {
        return getLimit() != 0 && enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public double getCostMultiplier() {
        return config.costMultiplier.get();
    }

    public void setCostMultiplier(double costMultiplier) {
        this.config.costMultiplier.set(costMultiplier);
    }

    public double getBonusMultiplier() {
        return config.bonusMultiplier.get();
    }

    public void setBonusMultiplier(double bonusMultiplier) {
        this.config.bonusMultiplier.set(bonusMultiplier);
    }
}
